bus-mapping/src/circuit_input_builder/execution.rs

```diff
@@ -1060,53 +1060,110 @@ impl EcMulOp {
 /// call are < 4, we append (G1::infinity, G2::generator) until we have the required no. of inputs.
 pub const N_PAIRING_PER_OP: usize = 4;
 
+/// The number of bytes taken to represent a pair (G1, G2).
+pub const N_BYTES_PER_PAIR: usize = 192;
+
+/// Pair of (G1, G2).
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct EcPairingPair {
+    /// G1 point.
+    pub g1_point: G1Affine,
+    /// G2 point.
+    pub g2_point: G2Affine,
+}
+
+impl EcPairingPair {
+    /// Returns the big-endian representation of the G1 point in the pair.
+    pub fn g1_bytes_be(&self) -> Vec<u8> {
+        std::iter::empty()
+            .chain(self.g1_point.x.to_bytes().iter().rev())
+            .chain(self.g1_point.y.to_bytes().iter().rev())
+            .cloned()
+            .collect()
+    }
+
+    /// Returns the big-endian representation of the G2 point in the pair.
+    pub fn g2_bytes_be(&self) -> Vec<u8> {
+        std::iter::empty()
+            .chain(self.g2_point.x.c0.to_bytes().iter().rev())
+            .chain(self.g2_point.x.c1.to_bytes().iter().rev())
+            .chain(self.g2_point.y.c0.to_bytes().iter().rev())
+            .chain(self.g2_point.y.c1.to_bytes().iter().rev())
+            .cloned()
+            .collect()
+    }
+
+    /// Returns the uncompressed big-endian byte representation of the (G1, G2) pair.
+    pub fn to_bytes_be(&self) -> Vec<u8> {
+        std::iter::empty()
+            .chain(self.g1_point.x.to_bytes().iter().rev())
+            .chain(self.g1_point.y.to_bytes().iter().rev())
+            .chain(self.g2_point.x.c0.to_bytes().iter().rev())
+            .chain(self.g2_point.x.c1.to_bytes().iter().rev())
+            .chain(self.g2_point.y.c0.to_bytes().iter().rev())
+            .chain(self.g2_point.y.c1.to_bytes().iter().rev())
+            .cloned()
+            .collect()
+    }
+
+    /// Create a new pair.
+    pub fn new(g1_point: G1Affine, g2_point: G2Affine) -> Self {
+        Self { g1_point, g2_point }
+    }
+
+    /// Padding pair for ECC circuit.
+    pub fn ecc_padding() -> Self {
+        Self {
+            g1_point: G1Affine::identity(),
+            g2_point: G2Affine::generator(),
+        }
+    }
+
+    /// Padding pair for EVM circuit.
+    pub fn evm_padding() -> Self {
```

Comment:
add some explanation here. Use (infinity, infinity) as evm padding since full 0 bytes is simpler. Use (identity, generator) as ecc padding since halo2-ecc does not support g2 as infinity now.